### PR TITLE
ENT-7502: Fixed Windows version unknown in MP

### PR DESCRIPTION
**Pull request by larsewi**
Tested:
 - [x] Windows 2012
 - [x] Windows 2016
 - [x] Windows 2019
 - [x] Windows 10

cherry-picked to 3.18.x - https://github.com/cfengine/core/pull/4762

Ticket: ENT-7502
Changelog: None
Signed-off-by: larsewi <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by larsewi** [comment]
@cf-bottom could you make me a windows build in Jenkins please ?

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6983/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6983/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6983/

**Comment by larsewi** [comment]
@cf-bottom Jenkins again please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6984/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6984/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6984/

File: libenv/sysinfo.c [hunk 1]
```diff
@@ -1554,7 +1560,18 @@ static void OSClasses(EvalContext *ctx)
         EvalContextClassPutHard(ctx, "unknown_ostype", "source=agent,derived-from=sys.sysname");
     }
 
-    SetFlavor(ctx, "windows");
+    char *release = SafeStringDuplicate(VSYSNAME.release);
+    char *major = NULL;
+    FindNextInteger(release, &major);
+    if NULL_OR_EMPTY(major)
```
Comment (olehermanse) [inline libenv/sysinfo.c:1566]:
```suggestion
    if (NULL_OR_EMPTY(major))
```

We are not savages ;)

**Review by olehermanse — CHANGES_REQUESTED** [review]
Could you test this manually on the various versions of windows? At least 2012, 2019, 10.

(Setting up a hub and installing and bootstrapping CFEngine on 3 different windows machines should be enough.

File: libenv/sysinfo.c [hunk 2]
```diff
@@ -150,6 +150,8 @@ static time_t GetBootTimeFromUptimeCommand(time_t now);
 
 /*****************************************************/
 
+static char *FindNextInteger(char *str, char **num);
+
```
Comment (olehermanse) [inline libenv/sysinfo.c:154]: This change means you have to update a line number in cppcheck_suppressions.txt:


```
// cppcheck is not clever enough to see that if (i >= PLATFORM_CONTEXT_MAX) then 'found' is false
arrayIndexOutOfBounds:libenv/sysinfo.c:585
```

**Review by olehermanse — APPROVED** [review]

File: libenv/sysinfo.c (hunk 2 above)
Comment (olehermanse) [inline libenv/sysinfo.c:154]: This change means you have to update a line number in cppcheck_suppressions.txt:


```
// cppcheck is not clever enough to see that if (i >= PLATFORM_CONTEXT_MAX) then 'found' is false
arrayIndexOutOfBounds:libenv/sysinfo.c:585
```

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: libenv/sysinfo.c (hunk 2 above)
Comment (larsewi) [reply libenv/sysinfo.c:154]: ✅

**Review by olehermanse — APPROVED** [review]

**Pull request merged**